### PR TITLE
[clang] Enable descriptions for --print-supported-extensions

### PR DESCRIPTION
**Pull request by cbalint13**
Enables summary descriptions along with the names of the feature.
Descriptions here are simply looked up via the available llvm tablegen data.

### CODE REVIEW

File: clang/tools/driver/cc1_main.cpp [hunk 1]
```diff
@@ -198,13 +199,21 @@ static int PrintSupportedExtensions(std::string TargetStr) {
   std::unique_ptr<llvm::TargetMachine> TheTargetMachine(
       TheTarget->createTargetMachine(TargetStr, "", "", Options, std::nullopt));
   const llvm::Triple &MachineTriple = TheTargetMachine->getTargetTriple();
+  const llvm::MCSubtargetInfo *MCInfo = TheTargetMachine->getMCSubtargetInfo();
+  const llvm::ArrayRef<llvm::SubtargetFeatureKV> Features =
+    MCInfo->getAllProcessorFeatures();
+
+  std::map<StringRef, StringRef> llvmDescMap;
```
Comment (DavidSpickett) [inline clang/tools/driver/cc1_main.cpp:206]: Potentially this could be https://llvm.org/doxygen/classllvm_1_1StringMap.html instead.

File: clang/tools/driver/cc1_main.cpp [hunk 2]
```diff
@@ -198,13 +199,21 @@ static int PrintSupportedExtensions(std::string TargetStr) {
...
+  for (const llvm::SubtargetFeatureKV &feature : Features) {
+    llvmDescMap.insert(std::make_pair(feature.Key, feature.Desc));
```
Comment (DavidSpickett) [inline clang/tools/driver/cc1_main.cpp:208]: https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

File: llvm/lib/Support/RISCVISAInfo.cpp [hunk 3]
```diff
@@ -210,24 +210,33 @@ static void verifyTables() {
...
   RISCVISAInfo::OrderedExtensionMap ExtMap;
   for (const auto &E : SupportedExtensions)
     ExtMap[E.Name] = {E.Version.Major, E.Version.Minor};
-  for (const auto &E : ExtMap)
-    outs() << format("\t%-20s%d.%d\n", E.first.c_str(), E.second.MajorVersion,
+  for (const auto &E : ExtMap) {
+    outs() << format("\t%-20s%d.%d", E.first.c_str(), E.second.MajorVersion,
                      E.second.MinorVersion);
+    outs() << (llvmDescMap.empty() ? "\n"
+                                   : ("\t\t" + llvmDescMap[E.first] + "\n"));
```
Comment (DavidSpickett) [inline llvm/lib/Support/RISCVISAInfo.cpp:226]: I think you might have an extra tab here, the risc-v output was misaligned.

File: llvm/lib/TargetParser/AArch64TargetParser.cpp [hunk 4]
```diff
@@ -11,11 +11,13 @@
 //
 //===----------------------------------------------------------------------===//
 
-#include "llvm/TargetParser/AArch64TargetParser.h"
+#include "llvm/Support/Error.h"
```
Comment (DavidSpickett) [inline llvm/lib/TargetParser/AArch64TargetParser.cpp:14]: Why do we need to include Support/Error now?

These includes shouldn't change too much, the "module" include `AArch64TargetParser.h` will always come first regardless of name order, clang-format should enforce that for you.

File: llvm/lib/TargetParser/AArch64TargetParser.cpp [hunk 5]
```diff
@@ -11,11 +11,13 @@
...
 #include "llvm/Support/raw_ostream.h"
+#include "llvm/TargetParser/AArch64TargetParser.h"
 #include "llvm/TargetParser/ARMTargetParserCommon.h"
 #include "llvm/TargetParser/Triple.h"
 #include <cctype>
+#include <map>
```
Comment (DavidSpickett) [inline llvm/lib/TargetParser/AArch64TargetParser.cpp:20]: Probably don't need to include map if `AArch64TargetParser.h` already references it.

File: llvm/lib/TargetParser/ARMTargetParser.cpp [hunk 6]
```diff
@@ -600,11 +601,17 @@ StringRef ARM::getARMCPUForArch(const llvm::Triple &Triple, StringRef MArch) {
...
   for (const auto &Ext : ARCHExtNames) {
     // Extensions without a feature cannot be used with -march.
-    if (!Ext.Feature.empty())
-      outs() << '\t' << Ext.Name << "\n";
+    if (!Ext.Feature.empty()) {
+      if (llvmDescMap.empty()) {
+        outs() << '\t' << Ext.Name << "\n";
+      } else {
+        outs() << format("\t%-20s", Ext.Name.str().c_str());
+        outs() << "\t\t" << llvmDescMap[Ext.Name] << "\n";
```
Comment (DavidSpickett) [inline llvm/lib/TargetParser/ARMTargetParser.cpp:613]: This can be one call to outs().

File: llvm/lib/TargetParser/ARMTargetParser.cpp [hunk 7]
```diff
@@ -600,11 +601,17 @@ StringRef ARM::getARMCPUForArch(const llvm::Triple &Triple, StringRef MArch) {
...
 }
 
-void ARM::PrintSupportedExtensions() {
+void ARM::PrintSupportedExtensions(std::map<StringRef, StringRef> llvmDescMap) {
   outs() << "All available -march extensions for ARM\n\n";
   for (const auto &Ext : ARCHExtNames) {
     // Extensions without a feature cannot be used with -march.
-    if (!Ext.Feature.empty())
-      outs() << '\t' << Ext.Name << "\n";
+    if (!Ext.Feature.empty()) {
```
Comment (DavidSpickett) [inline llvm/lib/TargetParser/ARMTargetParser.cpp:608]: Remove a level of indentation by making this:
```
if (Ext.Feature.empty())
  continue;

outs() << '\t' << Ext.Name << "\n";
<...>
```

Comment (DavidSpickett) [reply llvm/lib/TargetParser/ARMTargetParser.cpp:608]: https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code

File: llvm/unittests/TargetParser/TargetParserTest.cpp [hunk 8]
```diff
@@ -1016,7 +1017,8 @@ TEST(TargetParserTest, ARMPrintSupportedExtensions) {
 
   outs().flush();
   testing::internal::CaptureStdout();
-  ARM::PrintSupportedExtensions();
+  std::map<StringRef, StringRef> EmptyMap;
```
Comment (DavidSpickett) [inline llvm/unittests/TargetParser/TargetParserTest.cpp:1020]: All tests for this should include at least 1 description, "this is a description" is fine. Just to prove we don't crash.

File: llvm/lib/Support/RISCVISAInfo.cpp [hunk 9]
```diff
@@ -210,24 +210,33 @@ static void verifyTables() {
...
 
   RISCVISAInfo::OrderedExtensionMap ExtMap;
   for (const auto &E : SupportedExtensions)
     ExtMap[E.Name] = {E.Version.Major, E.Version.Minor};
-  for (const auto &E : ExtMap)
-    outs() << format("\t%-20s%d.%d\n", E.first.c_str(), E.second.MajorVersion,
+  for (const auto &E : ExtMap) {
+    outs() << format("\t%-20s%d.%d", E.first.c_str(), E.second.MajorVersion,
                      E.second.MinorVersion);
+    outs() << (llvmDescMap.empty() ? "\n"
```
Comment (DavidSpickett) [inline llvm/lib/Support/RISCVISAInfo.cpp:225]: In general I don't think we should assume that all named frontend features will be in the backend by the same name, and then also have a description.

Practically, that means I'd prefer we do a lookup for each name, expecting that it might fail. Instead of checking if the map is empty and if not doing an unconditional lookup.

In 99.9% of cases, the name probably matches. Maybe even has to match, but I know we've bent that rule in AArch64 in the past, and it wouldn't cost that much in terms of cpu time to do it that way. This is not a hot path, and we're spending time building a map already.

**Review by DavidSpickett — CHANGES_REQUESTED** [review]
This looks really nice. Two overall things:

Now that ARM and AArch64 print more than just names, they should have column headers like RISC-V does. Name / Description

The clang test case `clang/test/Driver/print-supported-extensions.c` should be extended to check at least the column names and 1 row of the extension details. This is because we're now combining a few functions to produce this output, so an integration test is needed.

It is a bit ugly to have to combine these 2 sources of information but that's a larger issue that shouldn't block this.

File: llvm/lib/Support/RISCVISAInfo.cpp (hunk 9 above)
Comment (cbalint13) [reply llvm/lib/Support/RISCVISAInfo.cpp:225]: > In general I don't think we should assume that all named frontend features will be in the backend by the same name, and then also have a description.

Yes, there are misses already, very few ones (less then five for all the 3 arches).
But this will encourage us to a better house-keeping in tablegen (backend) vs. here (frontend).

> Practically, that means I'd prefer we do a lookup for each name, expecting that it might fail. Instead of checking if the map is empty and if not doing an unconditional lookup.

The ```llvmDescMap.empty()``` is for other reason, I was thinking to make the whole ```Description``` column optional. It is useful for unit-test to bypass the whole target-machine creation saga and just look at the essential mandatory two columns (without descriptions column in test-mode).

* See in ```llvm/unittests/Support/RISCVISAInfoTest.cpp```
```
  std::map<StringRef, StringRef> EmptyMap;
  llvm::riscvExtensionsHelp(EmptyMap);
```
Basically with this feature we have an option to skip "Description" column at all.


> In 99.9% of cases, the name probably matches. Maybe even has to match, but I know we've bent that rule in AArch64 in the past, and it wouldn't cost that much in terms of cpu time to do it that way. This is not a hot path, and we're spending time building a map already.

The features amount per arch is less than a hundred, this is a user-help-menu anyway guiding the user.

File: llvm/lib/TargetParser/AArch64TargetParser.cpp (hunk 5 above)
Comment (cbalint13) [reply llvm/lib/TargetParser/AArch64TargetParser.cpp:20]: > Probably don't need to include map if `AArch64TargetParser.h` already references it.

It was included before, i just re-arrange in alphabetical (clang-format).

File: llvm/lib/TargetParser/ARMTargetParser.cpp (hunk 6 above)
Comment (cbalint13) [reply llvm/lib/TargetParser/ARMTargetParser.cpp:613]: > This can be one call to outs().

The ```if-else``` It is due to separation of "Description" column as optional.
To compress the 2*outs (inside else) format then will be quite long expansion: 
```llvmDescMap["experimental-"+Ext.Name].str().c_str() ```

Will try again, but spent quite some time on this very estetic part.

File: llvm/unittests/TargetParser/TargetParserTest.cpp (hunk 8 above)
Comment (cbalint13) [reply llvm/unittests/TargetParser/TargetParserTest.cpp:1020]: Will address this, even by adding a dummy description (to skip whole target-machine creation +feature-query).

**Comment by cbalint13** [comment]
@DavidSpickett 


Thank you for the review time, will address all the concerns and be back with a re-review request.

File: llvm/lib/TargetParser/ARMTargetParser.cpp (hunk 6 above)
Comment (DavidSpickett) [reply llvm/lib/TargetParser/ARMTargetParser.cpp:613]: Sorry, I think I wasn't clear, I meant this could be one call:
```
outs() << format("\t%-20s", Ext.Name.str().c_str());
outs() << "\t\t" << llvmDescMap[Ext.Name] << "\n";
```
As in:
```
outs() << format("\t%-20s", Ext.Name.str().c_str())
          << "\t\t" << llvmDescMap[Ext.Name] << "\n";
```

I didn't mean that having an if/else was a bad idea.

File: llvm/unittests/TargetParser/TargetParserTest.cpp (hunk 8 above)
Comment (DavidSpickett) [reply llvm/unittests/TargetParser/TargetParserTest.cpp:1020]: Yeah that's fine. All it should check is that there's no path where a lookup (successful or not) causes an exception.

File: llvm/lib/Support/RISCVISAInfo.cpp (hunk 9 above)
Comment (DavidSpickett) [reply llvm/lib/Support/RISCVISAInfo.cpp:225]: > Basically with this feature we have an option to skip "Description" column at all.

I see what you mean now. Yes I agree, if there's some architecture that provides 0 descriptions then having it doesn't make any sense.

**Comment by cbalint13** [comment]
@DavidSpickett ,


The new output: [console-output.txt.gz](https://github.com/llvm/llvm-project/files/12668489/console-output.txt.gz)
```
$ zcat console-output.txt.gz | grep 'print'
$ ./bin/clang --target=arm-unknown-linux-gnu --print-supported-extensions
$ ./bin/clang --target=aarch64-unknown-linux-gnu --print-supported-extensions
$ ./bin/clang --target=riscv64-unknown-linux-gnu --print-supported-extensions
```

Addressed all requests, let me know if still spot something.
Re-formatted the print without any '\t', but using spaces with precise justifiers.

File: clang/test/Driver/print-supported-extensions.c [hunk 10]
```diff
@@ -4,14 +4,17 @@
...
 
 // RUN: %if riscv-registered-target %{ %clang --target=riscv64-linux-gnu \
 // RUN:   --print-supported-extensions 2>&1 | FileCheck %s --check-prefix RISCV %}
 // RISCV: All available -march extensions for RISC-V
+// RISCV:     Name                Version   Description
 
 // RUN: %if arm-registered-target %{ %clang --target=arm-linux-gnu \
 // RUN:   --print-supported-extensions 2>&1 | FileCheck %s --check-prefix ARM %}
 // ARM: All available -march extensions for ARM
+// ARM:     Name                Description
```
Comment (DavidSpickett) [inline clang/test/Driver/print-supported-extensions.c:17]: Include the first line of the output as well, so we know it lines up (for all 3 architectures).

Don't use a `-NEXT:` check, just `ARM:` again. Since the extensions may get reordered later and we have no need to be strict.

File: llvm/unittests/TargetParser/TargetParserTest.cpp [hunk 11]
```diff
@@ -1012,11 +1013,18 @@ TEST(TargetParserTest, getARMCPUForArch) {
...
   std::string expected = "All available -march extensions for ARM\n\n"
-                         "\tcrc\n\tcrypto\n\tsha2";
+                         "    Name                Description\n"
+                         "    crc                 This is a long dummy description\n"
+                         "    crypto              This is a long dummy description\n"
+                         "    sha2                This is a long dummy description\n";
+
+  llvm::StringMap<llvm::StringRef> DummyMap;
+  for (const auto &E : llvm::ARM::ARCHExtNames)
+    DummyMap.insert({E.Name, "This is a long dummy description"});
```
Comment (DavidSpickett) [inline llvm/unittests/TargetParser/TargetParserTest.cpp:1023]: This is ok but what I meant specifically was, add a dummy description for just one extension. Then these tests are covering the case where we do and don't have an extension.

File: llvm/lib/TargetParser/ARMTargetParser.cpp [hunk 12]
```diff
@@ -600,11 +601,16 @@ StringRef ARM::getARMCPUForArch(const llvm::Triple &Triple, StringRef MArch) {
...
+         << "    " << left_justify("Name", 20)
+         << (llvmDescMap.empty() ? "\n" : "Description\n");
+
   for (const auto &Ext : ARCHExtNames) {
     // Extensions without a feature cannot be used with -march.
     if (!Ext.Feature.empty())
-      outs() << '\t' << Ext.Name << "\n";
+      outs() << format("    %-20s", Ext.Name.str().c_str())
+             << format(llvmDescMap.empty() ? "\n" : "%s\n",
+                       llvmDescMap[Ext.Name].str().c_str());
```
Comment (DavidSpickett) [inline llvm/lib/TargetParser/ARMTargetParser.cpp:614]: I was about to say this crashes if the lookup fails, it might in asserts builds, but given this is a map it'll default construct the value which in this case is empty string. Which is fine.

So leave this as is and ignore my earlier comments about the lookup, I forgot that maps have this quirk to them.

File: llvm/lib/Support/RISCVISAInfo.cpp [hunk 13]
```diff
@@ -210,24 +210,42 @@ static void verifyTables() {
...
-void llvm::riscvExtensionsHelp() {
-  outs() << "All available -march extensions for RISC-V\n\n";
-  outs() << '\t' << left_justify("Name", 20) << "Version\n";
+void llvm::getAllExtensions(StringMap<RISCVExtensionInfo> &ExtMap) {
+  for (const auto &E : SupportedExtensions)
+    ExtMap[E.Name] = {E.Version.Major, E.Version.Minor};
+  for (const auto &E : SupportedExperimentalExtensions)
+    ExtMap[(StringRef("experimental-") + E.Name).str().c_str()] = {
+        E.Version.Major, E.Version.Minor};
+}
```
Comment (DavidSpickett) [inline llvm/lib/Support/RISCVISAInfo.cpp:219]: Given that you don't need to add dummy descriptions to *all* extensions in the tests, this method becomes unused.

File: llvm/lib/Support/RISCVISAInfo.cpp [hunk 14]
```diff
@@ -210,24 +210,42 @@ static void verifyTables() {
...
-  for (const auto &E : ExtMap)
-    outs() << format("\t%-20s%d.%d\n", E.first.c_str(), E.second.MajorVersion,
-                     E.second.MinorVersion);
+  for (const auto &E : ExtMap) {
+    std::string Version = std::to_string(E.second.MajorVersion) + "." +
+                          std::to_string(E.second.MinorVersion);
+    outs() << format("    %-20s%-10s", E.first.c_str(), Version.c_str())
+           << format(llvmDescMap.empty() ? "\n" : "%s\n",
+                     llvmDescMap[E.first].str().c_str());
+  }
```
Comment (DavidSpickett) [inline llvm/lib/Support/RISCVISAInfo.cpp:236]: Given that this loop is almost the same for normal and experimental extensions, you could move this out into a static function `static void PrintExtensions` that takes a prefix that's either `""` or `"experimental-"`.

This would make it clear that both format in the same way, same padding etc.

File: llvm/lib/TargetParser/ARMTargetParser.cpp [hunk 15]
```diff
@@ -11,9 +11,10 @@
 //
 //===----------------------------------------------------------------------===//
 
-#include "llvm/TargetParser/ARMTargetParser.h"
 #include "llvm/ADT/StringSwitch.h"
+#include "llvm/Support/Format.h"
 #include "llvm/Support/raw_ostream.h"
+#include "llvm/TargetParser/ARMTargetParser.h"
```
Comment (DavidSpickett) [inline llvm/lib/TargetParser/ARMTargetParser.cpp:17]: Again, if you're manually re-ordering these please don't.

If clang-format is doing this, well that's a bit odd but I guess I'll yield to the robot on that one.

File: llvm/include/llvm/Support/RISCVISAInfo.h [hunk 16]
```diff
@@ -22,7 +23,8 @@ struct RISCVExtensionInfo {
   unsigned MinorVersion;
 };
 
-void riscvExtensionsHelp();
+void getAllExtensions(StringMap<RISCVExtensionInfo> &Features);
```
Comment (DavidSpickett) [inline llvm/include/llvm/Support/RISCVISAInfo.h:26]: This can go if you just set the description of only one extension in the tests.

File: clang/tools/driver/cc1_main.cpp [hunk 17]
```diff
@@ -198,13 +199,20 @@ static int PrintSupportedExtensions(std::string TargetStr) {
   std::unique_ptr<llvm::TargetMachine> TheTargetMachine(
       TheTarget->createTargetMachine(TargetStr, "", "", Options, std::nullopt));
   const llvm::Triple &MachineTriple = TheTargetMachine->getTargetTriple();
+  const llvm::MCSubtargetInfo *MCInfo = TheTargetMachine->getMCSubtargetInfo();
+  const llvm::ArrayRef<llvm::SubtargetFeatureKV> Features =
+    MCInfo->getAllProcessorFeatures();
+
+  llvm::StringMap<llvm::StringRef> llvmDescMap;
```
Comment (DavidSpickett) [inline clang/tools/driver/cc1_main.cpp:206]: I don't see the need for the `llvm` prefix here, I guess you're saying that these come from llvm as in the backend, which is true but not really useful information here.

`DescMap` would be fine instead.

File: clang/test/Driver/print-supported-extensions.c (hunk 10 above)
Comment (DavidSpickett) [reply clang/test/Driver/print-supported-extensions.c:17]: E.g.
```
ARM:     crc                 Enable support for CRC instructions
```
Then if we ever added `arc` it wouldn't matter if it came first, we're still checking the important things.

**Review by DavidSpickett — CHANGES_REQUESTED** [review]

File: clang/test/Driver/print-supported-extensions.c (hunk 10 above)
Comment (cbalint13) [reply clang/test/Driver/print-supported-extensions.c:17]: This was covered in unit-test already.

File: llvm/unittests/TargetParser/TargetParserTest.cpp (hunk 11 above)
Comment (cbalint13) [reply llvm/unittests/TargetParser/TargetParserTest.cpp:1023]: The way you tested RISCV previously in unittest (in full-test) require to fill up all fields.
I don't see harm to go for all fields.

File: llvm/include/llvm/Support/RISCVISAInfo.h (hunk 16 above)
Comment (cbalint13) [reply llvm/include/llvm/Support/RISCVISAInfo.h:26]: I would leave it.

It's fine for endusers to have an API with an accessor to the *full* list of what RISCV have.
Whole mine effort here started with lack of accessor to exactly such thing: https://github.com/llvm/llvm-project/pull/66582

File: llvm/unittests/TargetParser/TargetParserTest.cpp (hunk 11 above)
Comment (DavidSpickett) [reply llvm/unittests/TargetParser/TargetParserTest.cpp:1023]: The harm is we aren't testing what happens if an extension has no description. If the underlying implementation changes in the future to something that does fault on a missing description, no one will know until a user tries the option.

Comment (DavidSpickett) [reply llvm/unittests/TargetParser/TargetParserTest.cpp:1023]: Also it saves you adding a method to RISC-V just to get the list of extensions. Which we did for AArch64 but it was a hack there too that I would love to also remove one day.

File: clang/tools/driver/cc1_main.cpp (hunk 17 above)
Comment (cbalint13) [reply clang/tools/driver/cc1_main.cpp:206]: It indicate the origin from llvm backend.

File: clang/test/Driver/print-supported-extensions.c (hunk 10 above)
Comment (DavidSpickett) [reply clang/test/Driver/print-supported-extensions.c:17]: It is, and that's great, but, if someone decides to refactor this one day and suddenly they aren't using those functions anymore, we could miss it. It's unlikely but it's cheap to add a few lines here as well to cover that possibility.

Also, it's far easier for someone who is trying to understand the option to read this test file than it is to know that it calls 3 different methods in the background. So from a documentation point of view it also helps.

File: clang/tools/driver/cc1_main.cpp (hunk 17 above)
Comment (DavidSpickett) [reply clang/tools/driver/cc1_main.cpp:206]: Sure that's got some logic to it, but the code is clearly building a map with data from MCInfo, so it's implicit that it came from the llvm backend. And there isn't another source of data, so it's not like we have an llvm description map and a clang description map and need to know the difference.

Do as you wish it's not a big deal, but if you keep it I think the name should be `LLVMDescMap` to match https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly.

File: llvm/include/llvm/Support/RISCVISAInfo.h (hunk 16 above)
Comment (cbalint13) [reply llvm/include/llvm/Support/RISCVISAInfo.h:26]: Regarding accesors see [these lines]
(https://github.com/cbalint13/tvm/blob/4c17af1a187dee450ab89107df18623e8edadede/src/target/llvm/llvm_module.cc#L90-L107) in an external project.

--

The info desired there would be usefull in super simple queries like: "Is true that user desired CPU X have feature X", e.g. a concrete example, a user have a ```casadelake``` and check up if it is ```avx512bw``` capable, and also LLVM backend have it implemented or not (if it is older).

Comment (DavidSpickett) [reply llvm/include/llvm/Support/RISCVISAInfo.h:26]: Ok but I'll refer you to MaskRay's comment there:
```
We need justification to have this API, ideally tested by an in-tree target, otherwise this is prone to be removed by another contributor as an unused function.
```

There is no need to add a dummy description to every extension just for the test. Just pick one that already exists. Then there's no in tree use for this function and MaskRay's logic applies again.

The fact that you were able to add such a method without much trouble also tells us that carrying it in tree, unused, isn't going to save anyone much effort when it finally is needed.

Comment (DavidSpickett) [reply llvm/include/llvm/Support/RISCVISAInfo.h:26]: An external project needing something is not a reason to add methods in tree. It could be part of an argument to do so but in the context of this patch, this new RISC-V method is not needed and should be removed.

If you want to ease TVM's pain by upstreaming their changes you can do so, in a PR focused on that, with some in tree use case and testing.

File: clang/test/Driver/print-supported-extensions.c (hunk 10 above)
Comment (cbalint13) [reply clang/test/Driver/print-supported-extensions.c:17]: I'll add the extra checks.

File: clang/tools/driver/cc1_main.cpp (hunk 17 above)
Comment (cbalint13) [reply clang/tools/driver/cc1_main.cpp:206]: I'll leve it as sugessted ```DescMap```.

File: llvm/include/llvm/Support/RISCVISAInfo.h (hunk 16 above)
Comment (cbalint13) [reply llvm/include/llvm/Support/RISCVISAInfo.h:26]: It was men't as example. Anyway, prior certain versions (in our case 18) that accessor is still needed, won't go anywhere soon for compatibility.

Comment (cbalint13) [reply llvm/include/llvm/Support/RISCVISAInfo.h:26]: I remove it (all), and add only partial fillers.

File: llvm/lib/Support/RISCVISAInfo.cpp (hunk 14 above)
Comment (cbalint13) [reply llvm/lib/Support/RISCVISAInfo.cpp:236]: Not sure, does it worth ?

Comment (DavidSpickett) [reply llvm/lib/Support/RISCVISAInfo.cpp:236]: I would given that a reader has to parse the formatting specifiers, a few ifs and a lookup or two. If they only need to do that once that's nice.

But the two loops are visually identical apart from the "experimental-" so it's a close decision.

Go with what you think is best.

File: llvm/lib/TargetParser/ARMTargetParser.cpp (hunk 15 above)
Comment (cbalint13) [reply llvm/lib/TargetParser/ARMTargetParser.cpp:17]: Will revert.

File: llvm/unittests/TargetParser/TargetParserTest.cpp (hunk 11 above)
Comment (cbalint13) [reply llvm/unittests/TargetParser/TargetParserTest.cpp:1023]: As method, isn't straight possible in the way this RISCV class is now.

Comment (DavidSpickett) [reply llvm/unittests/TargetParser/TargetParserTest.cpp:1023]: Not sure what you mean.
```
DummyMap["crc"] = "This is a long dummy description";
```
Is what I am suggesting. Adjust the expected output as needed.

Do that for RISC-V and AArch64 too and there's no need for an extra method in RISC-V to get all the extension names.

File: clang/tools/driver/cc1_main.cpp [hunk 18]
```diff
@@ -198,13 +199,20 @@ static int PrintSupportedExtensions(std::string TargetStr) {
   std::unique_ptr<llvm::TargetMachine> TheTargetMachine(
       TheTarget->createTargetMachine(TargetStr, "", "", Options, std::nullopt));
   const llvm::Triple &MachineTriple = TheTargetMachine->getTargetTriple();
+  const llvm::MCSubtargetInfo *MCInfo = TheTargetMachine->getMCSubtargetInfo();
```
Comment (DavidSpickett) [inline clang/tools/driver/cc1_main.cpp:202]: I just thought of one more thing. What happens here if the build was configured without the target backend enabled?

For example if CMake was configred with `-DLLVM_TARGETS_TO_BUILD=ARM`, and you ask for the RISC-V extensions, does clang crash?

Though that is an unusual thing to do, but it is possible since clang knows about all possible triples even without a target backend. And besides that, clang shouldn't crash anyway. Simplest solution is to pass an empty description map I think, and add a test for that situation.

Comment (DavidSpickett) [reply clang/tools/driver/cc1_main.cpp:202]: It's possible `TheTargetMachine` is already null in this situation, and this is best fixed outside of this change.

I'll check that.

Comment (DavidSpickett) [reply clang/tools/driver/cc1_main.cpp:202]: I was wrong, clang doesn't recognise the triple if it wasn't enabled:
```
$ ./bin/clang --target=riscv64-linux-gnu --print-support
clang version 18.0.0 (https://github.com/llvm/llvm-project.git ebefe83c092e41d243829ab812bb650674e2f3d2)
Target: riscv64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/david.spickett/build-llvm-aarch64/./bin
No available targets are compatible with triple "riscv64-unknown-linux-gnu"
```

So we won't get to this code if the target backend is missing. No worries :)

**Comment by cbalint13** [comment]
@DavidSpickett ,

Addressed all requests.
* Here is the console output: [console-output.txt.gz](https://github.com/llvm/llvm-project/files/12677908/console-output.txt.gz)
```
$ zcat ~/console-output.txt.gz | grep 'print'
$ ./bin/clang --target=arm-unknown-linux-gnu --print-supported-extensions
$ ./bin/clang --target=aarch64-unknown-linux-gnu --print-supported-extensions
$ ./bin/clang --target=riscv64-unknown-linux-gnu --print-supported-extensions
```

File: clang/tools/driver/cc1_main.cpp (hunk 18 above)
Comment (cbalint13) [reply clang/tools/driver/cc1_main.cpp:202]: On my side on this topic I already checked this situation, yes looks protected well.

@DavidSpickett 

As sidenote here, on this topic opened by you: 

* In case of ```createTargetMachine```, notice the -mcpu="" (second argument), it will force to **return all target information**. If you specify i.e. "sandybridge" than you can only access list of features and descriptions only related to "sandybridge" subarchtecture / subset .

* I really hope LLVM will keep this (-mcpu="" => ALL) , and not cut it out (some developer will think is a "safer" API), otherwise it will be impossible to query LLVM features of **a whole target**.

LLVM looks (to humble me) very dynamic with frequent API changes, my point here is a "utilitarian view" as external user I would like to take benefits and features through API.

**Review by DavidSpickett — APPROVED** [review]

**Comment by DavidSpickett** [comment]
Looks good to me, thanks for working on this.

I'm not sure if you'll have permissions to click the merge button but if you do, go ahead. Please remember to edit the final commit message to remove the links to the output files.

**Comment by cbalint13** [comment]
@DavidSpickett ,

> Looks good to me, thanks for working on this.

Thank you very much for the time !

> I'm not sure if you'll have permissions to click the merge button but if you do, go ahead. 

* No, don't, I am outside contributor.

>Please remember to edit the  final commit message to remove the links to the output files.

* Done.

**Comment by DavidSpickett** [comment]
If you are doing more contributions you can get commit access by doing https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access. Though with the switch to Github it's super easy to ask the reviewer to click the merge button, so it's not required.
